### ISSUE TITLE
Disconnections from Monti cause CPU increase / memory leak

### ISSUE THREAD

**Issue by bcluyse** [monti-apm/monti-apm-agent#37]
Our application had connectivity issues which resulted in the following errors approximately 50 times in total:
Monti APM Error: Reached maximum retry limit for Request failed: 502
Monti APM Error: Reached maximum retry limit for Request failed: 520
Monti APM Error: Reached maximum retry limit for Request failed: 525
and only after starting a new instance, we started getting statistics again on the dashboard page.

![image](https://user-images.githubusercontent.com/20240779/106466213-585b0280-649b-11eb-9846-99f6f5fb878a.png)

However, the instances that are still alive have been slowly increasing CPU and memory (see below) and after taking a CPU profile we verified that quite some time is spent in Monti internals. (I cannot share the profiles because of company policy).
Up to 4s is spent in one iteration of TracerStore.processTraces which does not really sound reasonable.

Details:
Meteor 1.11
Monti Version: 2.39.1

![image](https://user-images.githubusercontent.com/20240779/106466873-2dbd7980-649c-11eb-828e-2639186b8cc8.png)

**Comment by zodern** [comment]
Please try updating to the latest version (2.40.1). `TracerStore.processTraces` is much more efficient now. The agent does have some caches, so during the first few hours the app runs memory will grow until the caches are filled, but it looks like there is a memory leak in your app. I will look at the tracer store this week to make sure it isn't the cause, and look into the connection issues with your app.

**Comment by bcluyse** [comment]
Thanks for the quick update, we indeed seem to still have a memory leak in our application so we are investigating that further. The root cause is for sure not Monti, if there is however any trace from Monti side on the initial disconnects that might point already in a certain direction.

**Comment by zodern** [comment]
This week I improved how we are monitoring the API used by the agent, and fixed many causes behind requests failing. The service should be much more reliable now, and we will be notified when new issues appear.

I will check if the agent has a memory leak next week.

**Comment by bcluyse** [comment]
The version I originally posted is incorrect, the running version is with 2.39.0-beta7.
Could this somehow be related to the memory issues we have been seeing?

**Comment by zodern** [comment]
I don't think the changes in the recent versions of the agent should affect the memory usage.

What are you using to deploy your app? If you are using Meteor Up, you can run `mup meteor debug` to use Chrome's Devtools for Node to take heap snapshots to analyze the memory leak.

**Comment by bcluyse** [comment]
So, we disabled Monti by changing the environment variables to invalid keys (which was not a good idea).
Eventually, we were able to get heap snapshots from production and it seems that the TracerStore.traceArchives were still filling up. Now, we have completely removed Monti from the image for the time being and the memory issues have been completely resolved.  So, the initial memory leak probably was probably triggered after the initial 'disconnect'.

So it comes down to the following 2 issues
- We did not expect the package to keep on filling the archive even when not successfully connecting (but this is not really a problem since we should disable it in other ways)
- We need some way to prevent this from happening in the future, these disconnections are not really problematic but due to the nature of our app we cannot restart the containers on-the-fly if the same thing happens as initially and memory+cpu starts getting out of control.

**Comment by zodern** [comment]
I published version 2.44.1 of the agent to fix the memory leak when it is not able to connect.

The memory leak caused by the agent disconnecting should have been fixed in 2.39.0-beta8.

**Comment by claudiovicar** [comment]
Hey folks, we just had this exact same issue today with version `2.45.2-beta.2`.
Our servers were forced to restart due to a memory error some time after a Monti connection failure.

`2022-06-21T06:30:55.965598170Z Monti APM Error: while sending payload to Monti APM: Reached maximum retry limit for Request failed: 500`

Then, around 10 minutes later (the error kept happening in the meantime):
```
2022-06-21T06:42:36.417383217Z Monti APM Error: while sending payload to Monti APM: Reached maximum retry limit for Request failed: 500
2022-06-21T06:42:53.332755460Z <--- JS stacktrace --->
2022-06-21T06:42:53.332760121Z 
2022-06-21T06:42:53.332764375Z FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
```

Could it be related to this issue? Any help is appreciated!
Thanks in advance
